[IDF] Allow `add_idf_component` in PR (clang-tidy include error)

### The problem

With ESP-IDF, if we submit a PR for a component that uses `add_idf_component(..)` in `__init__.py` and includes some files from this IDF component, the continuous integration  for `clang-tidy for ESP32 IDF` fails because the IDF component is not installed before clang-tidy runs.

Using `// NOLINT` on the `#include` directive has no effect.

### Which version of ESPHome has the issue?

2024.5.4

### What type of installation are you using?

pip

### Which version of Home Assistant has the issue?

_No response_

### What platform are you using?

ESP32-IDF

### Board

_No response_

### Component causing the issue

_No response_

### Example YAML snippet

_No response_

### Anything in the logs that might be useful for us?

```txt
Error: /home/runner/work/esphome/esphome/esphome/components/modem/modem_component.h:19:10: error: 'cxx_include/esp_modem_api.hpp' file not found [clang-diagnostic-error]
#include <cxx_include/esp_modem_api.hpp>  // NOLINT
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```


### Additional information

_No response_

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -387,6 +387,13 @@ jobs:
           echo "::add-matcher::.github/workflows/matchers/gcc.json"
           echo "::add-matcher::.github/workflows/matchers/clang-tidy.json"
 
+      - name: Run 'pio run --list-targets -e esp32-idf-tidy'
+        if: matrix.name == 'Run script/clang-tidy for ESP32 IDF'
+        run: |
+          . venv/bin/activate
+          mkdir -p .temp
+          pio run --list-targets -e esp32-idf-tidy
+
       - name: Run clang-tidy
         run: |
           . venv/bin/activate

--- a/esphome/idf_component.yml
+++ b/esphome/idf_component.yml
@@ -1,0 +1,10 @@
+dependencies:
+  esp32_camera:
+    git: https://github.com/espressif/esp32-camera.git
+    version: v2.0.9
+  mdns:
+    git: https://github.com/espressif/esp-protocols.git
+    version: mdns-v1.2.5
+    path: components/mdns
+    rules:
+      - if: "idf_version >=5.0"

--- a/esphome/idf_component.yml
+++ b/esphome/idf_component.yml
@@ -1,7 +1,6 @@
 dependencies:
-  esp32_camera:
-    git: https://github.com/espressif/esp32-camera.git
-    version: v2.0.9
+  esp-tflite-micro:
+    git: https://github.com/espressif/esp-tflite-micro.git
   mdns:
     git: https://github.com/espressif/esp-protocols.git
     version: mdns-v1.2.5

--- a/script/clang-tidy
+++ b/script/clang-tidy
@@ -101,8 +101,10 @@ def clang_options(idedata):
     # add library include directories using -isystem to suppress their errors
     for directory in sorted(set(idedata["includes"]["build"])):
         # skip our own directories, we add those later
-        if not directory.startswith(f"{root_path}/") or directory.startswith(
-            f"{root_path}/.pio/"
+        if (
+            not directory.startswith(f"{root_path}/")
+            or directory.startswith(f"{root_path}/.pio/")
+            or directory.startswith(f"{root_path}/managed_components/")
         ):
             cmd.extend(["-isystem", directory])
 

